documentation/general/dash-sai-pipeline-packet-flow.md

```diff
@@ -0,0 +1,886 @@
+
+At a high level the pipeline looks like below:
+
+```mermaid
+flowchart TB
+    PP[Packet<br>Parsing]
+    DL[Direction<br>Lookup]
+    PL[Pipeline<br>Lookup]
+    PD[Packet<br>Decap]
+
+    subgraph P0[DASH Pipeline 0]
+        subgraph In0[Inbound Pipeline]
+            CTL0[Conntrack<br>Lookup]
+            PreACL0[Pre-pipeline<br>ACL]
+            MATS0[Matching<br>Stages]
+            AA0[Action<br>Apply]
+            PostACL0[Post-pipeline<br>ACL]
+            CTU0[Conntrack<br>Update]
+            MS0[Meter<br>Update]
+        end
+
+        subgraph Out0[Outbound Pipeline]
+            CTL1[Conntrack<br>Lookup]
+            PreACL1[Pre-pipeline<br>ACL]
+            MATS1[Matching<br>Stages]
+            AA1[Action<br>Apply]
+            PostACL1[Post-pipeline<br>ACL]
+            CTU1[Conntrack<br>Update]
+            MS1[Meter<br>Update]
+        end
+    end
+
+    subgraph P1[DASH Pipeline 1]
+        In1[Inbound Pipeline<br>...]
+        Out1[Outbound Pipeline<br>...]
+    end
+
+    Out[Packet<br>Out]
+
+    PP --> DL
+    DL --> PL
+    PL --> PD
+
+    PD --> In0
+    PD --> Out0
+    PD --> In1
+    PD --> Out1 
+    PD --> |No Pipeline<br>Match| Out
+
+    CTL0 --> |Flow Miss| PreACL0
+    CTL0 --> |Flow Hit| AA0
+    PreACL0 --> MATS0
+    MATS0 --> AA0
+    AA0 --> |Flow Miss| PostACL0
+    PostACL0 --> CTU0
+    AA0 --> |Flow Hit| MS0
+    AA0 --> |Flow Change| CTU0
+    CTU0 --> MS0
+    MS0 --> Out
+
+    CTL1 --> |Flow Miss| PreACL1
+    CTL1 --> |Flow Hit| AA1
+    PreACL1 --> MATS1
+    MATS1 --> AA1
+    AA1 --> |Flow Miss| PostACL1
+    PostACL1 --> CTU1
+    AA1 --> |Flow Hit| MS1
+    AA1 --> |Flow Change| CTU1
+    CTU1 --> MS1
+    MS1 --> Out
+```
+
+This design allows our upper layers to be flexible and doesn't limit to any specific object models for modeling their own policies. For example, [SONiC-DASH pipeline](https://github.com/sonic-net/SONiC/blob/master/doc/dash/dash-sonic-hld.md#2-packet-flows) uses DASH-SAI pipeline to implement VNET routing scenarios by translating the [SONiC DASH APIs](https://github.com/sonic-net/sonic-dash-api/tree/master/proto) to DASH-SAI model:
+
+- A DASH-SAI pipeline is used to represent a VM NIC (ENI).
+- The VxLAN VNI is used to do the direction lookup.
+- The inner MAC address is used for pipeline lookup, a.k.a. ENI lookup (ENI).
+- Once it enters the corresponding DASH pipeline, the `outbound` pipeline will be used to process the packets sourced from the VM, while the `inbound` pipeline will be used to process the packets destined to the VM.
+
+## 5. Pipeline components
+
+### 5.1. Per-packet metadata bus
+
+First of all, since we have multiple matching stages in the pipeline, we need a way to pass the information from the matched entries in the earlier stages to the later ones to help us making final decisions on packet transformation. And this is what metadata bus is for.
```

Comment:
```suggestion
First of all, since we have multiple matching stages in the pipeline, we need a way to pass the information from the matched entries in the earlier stages to the later ones to help us make final decisions on packet transformation. And this is what metadata bus is for.
```